probe-rs/src/probe.rs

```diff
@@ -794,63 +799,603 @@ impl PartialEq for dyn ProbeFactory {
+            && hid_interface_matches
+            && sysfs_path_matches
+            && instance_id_matches
+            && parent_instance_id_matches
+            && port_number_matches
+            && driver_matches
+            && registry_id_matches
+            && location_id_matches
+    }
+}
+
+macro_rules! make_filter_impls {
+    (
+        $ty:ty: [
+            $(
+                $(#[$($meta:tt)+])?
+                $method:ident$([$($generics:tt)+])?($field:ident: $fieldty:tt$(, [$($do:tt)+])?)
+            ),+
+            $(,)?
+        ]
+    ) => {
+        impl $ty {
+            $(
+                make_filter_impls! {
+                    @method
+                    meta: [$($($meta)+)?],
+                    info: $method$([$($generics)+])?($field: $fieldty$(, [$($do)+])?),
+                }
+            )+
+        }
+    };
+    (
+        @method
+        meta: [cfg($($cfg:tt)+)],
+        info: $method:ident$([$($generics:tt)+])?($field:ident: $fieldty:tt$(, [$($do:tt)+])?),
+    ) => {
+        #[cfg($($cfg)+)]
+        #[cfg_attr(
+            $($cfg)+,
+            doc = concat!("Sets the `", stringify!($field), "` field of a `UsbFilters` instance."),
+        )]
+        pub fn $method$(<$($generics)+>)?(mut self, $field: $fieldty) -> Self {
+            self.$field = Some($field$($($do)+)?);
+            self
+        }
+    };
+    (
+        @method
+        meta: [$($($other:tt)+)?],
+        info: $method:ident$([$($generics:tt)+])?($field:ident: $fieldty:tt$(, [$($do:tt)+])?),
+    ) => {
+        $(#[$($other)+])?
+        #[doc = concat!("Sets the `", stringify!($field), "` field of a `UsbFilters` instance.")]
+        pub fn $method$(<$($generics)+>)?(mut self, $field: $fieldty) -> Self {
+            self.$field = Some($field$($($do)+)?);
+            self
+        }
+    };
+}
+
+make_filter_impls! {
+    UsbFilters: [
+        with_serial_number[S: Into<String>](serial_number: S, [.into()]),
+        with_hid_interface(hid_interface: u8),
+
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        with_sysfs_path[P: AsRef<Path>](sysfs_path: P, [.as_ref().to_path_buf()]),
+
+        #[cfg(target_os = "windows")]
+        with_instance_id[S: Into<String>](instance_id: S, [.into()]),
+        #[cfg(target_os = "windows")]
+        with_parent_instance_id[S: Into<String>](parent_instance_id: S, [.into()]),
+        #[cfg(target_os = "windows")]
+        with_port_number(port_number: u32),
+        #[cfg(target_os = "windows")]
+        with_driver[S: Into<String>](driver: S, [.into()]),
+
+        #[cfg(target_os = "macos")]
+        with_registry_id(registry_id: u64),
+        #[cfg(target_os = "macos")]
+        with_location_id(location_id: u32),
+    ]
+}
+
+trait OptionFilter<T> {
+    fn filter_option<U>(&self, val: U) -> bool
+    where
+        for<'a> &'a T: PartialEq<U>;
+    fn filter_optional<U>(&self, val: Option<U>) -> bool
+    where
+        for<'a> &'a T: PartialEq<U>;
```

Comment:
You can do a slight trick to cut down this mess to ~half of it. Make your trait methods return `Option<()>`, where Some(()) represents true, and None represents false. Then define an inner function that has your comparison logic and returns Option<()>, where you can just write `xy.filter_option(z)?;` and end up with a short-circuiting condition like you have now, without declaring temporaries and writing a bit unnecessary and at the end.

In the end your matches will be:

```rust
fn matches(...) {
    inner_matches(...).is_some()
}
```

Besides, the pair of `filter_option`+`filter_optional` is rather suboptimal, make those names a bit more visually distinct please, e.g. by dropping `_option` from the first one.